src/main/scala/com/fulcrumgenomics/util/ReadStructure.scala

```diff
@@ -0,0 +1,143 @@
+    readStructure.filter(s => s.symbol == symbol.getOrElse(s.symbol) && (strict || s.offset < bases.length)).map {
+      segment =>
+        val end = segment.offset + segment.length
+        if (bases.length < end) {
+          if (strict) throw new IllegalStateException("Not enough bases for read structure: " + readStructure)
+          else (bases.substring(segment.offset), ReadSegment(segment, bases.length-segment.offset))
+        }
+        else (bases.substring(segment.offset, end), segment)
+    }
+  }
+
+  /** Splits the given bases and qualities into triples with its associated read segment. */
+  def extractWithQualities(bases: String, qualities: String, readStructure: ReadStructure, symbol: Option[BaseSymbol] = None, strict: Boolean = true): Seq[(String, String, ReadSegment)] = {
+    assert(bases.length == qualities.length)
+    readStructure.filter(s => s.symbol == symbol.getOrElse(s.symbol) && (strict || s.offset < bases.length)).map {
+      segment =>
+        val end = segment.offset + segment.length
+        if (strict) {
+          if (bases.length < end) throw new IllegalStateException("Not enough bases for read structure: " + readStructure)
+          if (qualities.length < end) throw new IllegalStateException("Not enough qualities for read structure: " + readStructure)
+        }
+        if (bases.length < end) (bases.substring(segment.offset), qualities.substring(segment.offset), ReadSegment(segment, bases.length-segment.offset))
+        else (bases.substring(segment.offset, end), qualities.substring(segment.offset, end), segment)
+    }
+  }
+
+  def template(rs: ReadStructure): Seq[Template] = rs.filter(_.symbol == TemplateSymbol).map(_.asInstanceOf[Template])
+  def sampleBarcode(rs: ReadStructure): Seq[SampleBarcode] = rs.filter(_.symbol == SampleBarcodeSymbol).map(_.asInstanceOf[SampleBarcode])
+  def molecularBarcode(rs: ReadStructure): Seq[MolecularBarcode] = rs.filter(_.symbol == MolecularBarcodeSymbol).map(_.asInstanceOf[MolecularBarcode])
+  def skip(rs: ReadStructure): Seq[Skip] = rs.filter(_.symbol == SkipSymbol).map(_.asInstanceOf[Skip])
+}
+
+/**
+  * Describes the structure of a give read.  A read contains one or more read segments. A read segment describes
+  * a contiguous stretch of bases of the same type (ex. template bases) of some length and some offset from the start
+  * of the read.
+  */
+class ReadStructure private(segs: Traversable[ReadSegment]) extends scala.collection.immutable.Seq[ReadSegment] {
+  private val segments: Seq[ReadSegment] = segs.toSeq
+  def length: Int = segments.length
+  def apply(idx: Int): ReadSegment = segments(idx)
+  def iterator: Iterator[ReadSegment] = segments.iterator
+  override def toString: String = segments.map(_.toString).mkString
+  def totalBases: Int = segments.map(_.length).sum
+}
+
+object BaseSymbol {
+  sealed trait BaseSymbol {
```

Comment:
I think it's weird to have a trait nested inside an object of the same name.  E.g. I would call the object `BaseSymbols` in this case.

But also, I think the name is kind of weird.  These are really describing the types of kinds of read segments.  They _contain_ the symbol used for that type of segment.  And other than being single characters don't, I think, resemble bases.  How about `SegmentType` instead?  And then I would suggest the case objects could be just `Template`, `SampleBarcode`, `MolecularBarcode`, `Skip` and reference them elsewhere as `SegmentType.Template` etc.